core/client/src/main/java/alluxio/client/file/FileSystemWorkerClient.java

```diff
@@ -0,0 +1,279 @@
+ * guarantees thread safety.
+ */
+// TODO(calvin): Session logic can be abstracted
+@ThreadSafe
+public class FileSystemWorkerClient extends AbstractClient {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /** Executor service for running the heartbeat thread. */
+  private final ExecutorService mExecutorService;
+  /** Heartbeat executor for the session heartbeat thread. */
+  private final HeartbeatExecutor mHeartbeatExecutor;
+  /** Metrics object. */
+  private final ClientMetrics mClientMetrics;
+  /** Address of the data server on the worker. */
+  private final InetSocketAddress mWorkerDataServerAddress;
+
+  /** Underlying thrift RPC client which executes the operations. */
+  private FileSystemWorkerClientService.Client mClient;
+  /** The current heartbeat thread, this will be updated each time this client connects. */
+  private Future<?> mHeartbeat;
+  /** The current session id, managed by the caller. */
+  private long mSessionId;
+
+  /**
+   * Constructor for a client that communicates with the {@link FileSystemWorkerClientService}.
+   *
+   * @param workerNetAddress the worker address to connect to
+   * @param executorService the executor service to run this client's heartbeat thread
+   * @param conf the configuration to use
+   * @param sessionId the session id to use, this should be unique
+   * @param metrics the metrics object to send any metrics through
+   */
+  public FileSystemWorkerClient(WorkerNetAddress workerNetAddress, ExecutorService executorService,
+      Configuration conf, long sessionId, ClientMetrics metrics) {
+    super(NetworkAddressUtils.getRpcPortSocketAddress(workerNetAddress), conf, "FileSystemWorker");
+    mWorkerDataServerAddress = NetworkAddressUtils.getDataPortSocketAddress(workerNetAddress);
+    mExecutorService = Preconditions.checkNotNull(executorService);
+    mSessionId = sessionId;
+    mClientMetrics = Preconditions.checkNotNull(metrics);
+    mHeartbeatExecutor = new FileSystemWorkerClientHeartbeatExecutor(this);
+  }
+
+  /**
+   * Creates the underlying thrift client and starts the heartbeat thread, unless one is already
+   * running.
+   *
+   * @throws IOException if the thrift client cannot be created
+   */
+  @Override
+  protected synchronized void afterConnect() throws IOException {
+    mClient = new FileSystemWorkerClientService.Client(mProtocol);
+    // only start the heartbeat thread if the connection is successful and if there is not
+    // another heartbeat thread running
+    if (mHeartbeat == null || mHeartbeat.isCancelled() || mHeartbeat.isDone()) {
+      final int interval = mConfiguration.getInt(Constants.USER_HEARTBEAT_INTERVAL_MS);
+      mHeartbeat =
+          mExecutorService.submit(new HeartbeatThread(HeartbeatContext.WORKER_CLIENT,
+              mHeartbeatExecutor, interval));
+    }
+  }
+
+  /**
+   * Cancels the current heartbeat thread.
+   */
+  @Override
+  protected synchronized void afterDisconnect() {
```

Comment:
why we have a bunch of `protected` methods in here? Do we expect them to be inherited?